File: 0013-roman-to-integer/0013-roman-to-integer.py
class Solution:
    def romanToInt(self, s):
        # Define a map for Roman numeral values
        roman_map = {
            'I': 1, 'V': 5, 'X': 10, 'L': 50,
            'C': 100, 'D': 500, 'M': 1000
        }
        
        total = 0  # To store the result
        prev_value = 0  # To keep track of the previous numeral's value
        
        # Process each Roman numeral in reverse order
        for char in reversed(s):
            current_value = roman_map[char]
            
            # If the current value is less than the previous, subtract it (subtractive notation)
            if current_value < prev_value:
                total -= current_value
            else:
                total += current_value
            
            # Update the previous value
            prev_value = current_value

        return total

# Example usage
if __name__ == "__main__":
    solution = Solution()
    print(solution.romanToInt("MMMDCCXLIX"))  # Output: 3749
    print(solution.romanToInt("LVIII"))       # Output: 58
    print(solution.romanToInt("MCMXCIV"))     # Output: 1994

File: 0013-roman-to-integer/0013-roman-to-integer.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int romanToInt(String s) {
        // Map to store Roman numeral values
        Map<Character, Integer> romanMap = new HashMap<>();
        romanMap.put('I', 1);
        romanMap.put('V', 5);
        romanMap.put('X', 10);
        romanMap.put('L', 50);
        romanMap.put('C', 100);
        romanMap.put('D', 500);
        romanMap.put('M', 1000);

        int result = 0;
        int prevValue = 0;

        // Traverse the string from right to left
        for (int i = s.length() - 1; i >= 0; i--) {
            char currentChar = s.charAt(i);
            int currentValue = romanMap.get(currentChar);

            // If the current value is less than the previous value, subtract it
            if (currentValue < prevValue) {
                result -= currentValue;
            } else { // Otherwise, add it
                result += currentValue;
            }

            // Update previous value
            prevValue = currentValue;
        }

        return result;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test cases
        System.out.println(solution.romanToInt("III"));       // Output: 3
        System.out.println(solution.romanToInt("LVIII"));     // Output: 58
        System.out.println(solution.romanToInt("MCMXCIV"));   // Output: 1994
    }
}
